### DISCUSSION TITLE
Volume Cartographer modules for Python

### DISCUSSION THREAD

**Discussion by csparker247** [educelab/volume-cartographer#10] [Ideas]
When we started Volume Cartographer, we selected C++ as our primary API because we thought that virtual unwrapping would benefit from object-oriented class structures, high-performance, and low-level memory management. We still think some of those things could be useful, but we could never have anticipated the stunning growth of machine learning frameworks and their heavy reliance on Python.

We're looking for contributors who will help us bring Volume Cartographer and Python closer together so that developers can more easily apply machine learning to the virtual unwrapping task. We already have experimental Python bindings, but these are just that: experimental. We would love someone to help us make the Python version of Volume Cartographer a first-class experience.

This discussion is to figure out the direction a Python interface should take. A few questions that immediately come to mind are: How do we do bindings across all of our libraries? What approach do we take ([pybind11](https://github.com/pybind/pybind11), [Shiboken](https://doc.qt.io/qtforpython/shiboken6/index.html), etc.)? What needs to change in our existing C++ code to make the transition easier?

Feel free to add your voice to the discussion!

**Comment by paramsingh** [comment]
>How do we do bindings across all of our libraries? 

Sorry if this is a dumb question, but could you list what libraries you envision? I would appreciate a more detailed list of things that are not present (or not good) in the current bindings that we want to improve.

(context: i have python experience but very little ML/DL experience, I'm wondering if I can contribute here somehow).

**Reply by csparker247** [reply]
The Volume Cartographer project is already comprised of a number of C++ libraries: core, meshing, segmentation, and texturing primarily. Each of these have their own directories in the root directory. 

The core library has some example bindings created using pybind11. These were experimental and just to test out how difficult it would be to instrument the whole project. Going forward, we want to pick a binding method that's easy to maintain. That may or may not be pybind11. 

Practically, we probably want to start by instrumenting the core library. It has all the basic types and utilities that all of the other libraries need.
